### PR TITLE
Refactor http module

### PR DESCRIPTION
**Pull request by anuptalwalkar**
Middlewares are defined on router than the handler. Now they can be defined and build separately from handlers. Following to this change, we want to separate and build middlewares specific to the defined routes.

### CODE REVIEW

**Comment by blampe** [comment]
Let's chat about uhttp before landing this.

cc @breerly @akshayjshah @glibsm @madhuravi

File: modules/uhttp/router.go [hunk 1]
```diff
@@ -22,17 +22,42 @@ package uhttp
...
+	}
+}
+
+type route struct {
+	matches RequestMatcher
+	handler http.Handler
+}
+
 // Router is wrapper around gorila mux
 type Router struct {
```
Comment (alsamylkin) [inline modules/uhttp/router.go]: I don't think we should have a router/custom interfaces in the http module. All we need to do is accept http.Handler interface and attach some metrics to it, then users can attach any routing frameworks they want.

Comment (anuptalwalkar) [inline modules/uhttp/router.go]: the whole point of having the wrapper is to inject middlewares that we have, not necessarily some metrics, but additional libraries such as rate limiting, etc. current mux.Router doesn't have the middleware chaining functionality.

Comment (alsamylkin) [inline modules/uhttp/router.go]: Rate limiter can also implement http.Handler, so we can chain middleware.

Comment (anuptalwalkar) [inline modules/uhttp/router.go]: I am confused. we are chaining using http.Handler only, except http.Handler doesn't have chaining. Thats why we have middleware to provide the chaining. you can't chain handlers with implementing only http.Handler as is.

Comment (alsamylkin) [inline modules/uhttp/router.go]: Will this work?
```golang
type AlexMiddleware struct {
	next http.Handler
}

func (a *AlexMiddleware) ServeHTTP(w ResponseWriter, rq *Request) {
	fmt.Println("Alex was here")
	a.next.ServeHTTP(w, rq)
}
```

**Pull request closed without merging**